### DISCUSSION TITLE
Enhance type matching

### DISCUSSION THREAD

**Discussion by ahdung** [dotnet/csharplang#5565] [Language Ideas]
I hope can do this:
```c#
void Foo(Type type) {
if (type is int or Array or Delegate or IEnumerable or IEnumerable<> or IEnumerable<string>)
    ...
}
```
and this:
```c#
void Foo<T>() {
if (T is int or Array or Delegate or IEnumerable or IEnumerable<> or IEnumerable<string>)
    ...
}
```

**Comment by huoyaoyuan** [comment]
You can already do this with `typeof(T)`. The really interesting thing is `if (T is IFloatingPoint<T>)` etc.

**Reply by DavidArno** [reply]
It's not currently possible to use `typeof` with patterns (other than in a when guard) as `typeof` isn't a constant. 

The lack of the ability to match `Type` values directly in patterns as the OP suggests is a big hole in C#'s pattern matching abilities. But no one in the language team seems interested in championing this feature, which is a huge shame.

**Reply by CyrusNajmabadi** [reply]
> suggests is a big hole in

I've honestly never run into it.  It doesn't really feel like a big hole.  More just something that might arise in some reflection cases.  And, in the above, i genuinely don't see why just using the reflection api would be bad.  for example `type == typeof(int) || ...`

**Reply by jnm2** [reply]
I run into this very rarely, and it's annoyingly verbose code to write every time.

Here's the equivalent code from the original post:

```cs
if (typeof(T) == typeof(int)
    || typeof(Array).IsAssignableFrom(typeof(T))
    || typeof(Delegate).IsAssignableFrom(typeof(T))
    || typeof(IEnumerable).IsAssignableFrom(typeof(T))
    || (typeof(T).IsGenericType && typeof(T).GetGenericTypeDefinition() == typeof(IEnumerable<>))
    || typeof(T).GetInterfaces().Any(interfaceType => interfaceType.IsGenericType && interfaceType.GetGenericTypeDefinition() == typeof(IEnumerable<>)))
    || typeof(IEnumerable<string>).IsAssignableFrom(typeof(T)))
```

**Reply by msedi** [reply]
I'm also running into this very often in many different situations:

- Plugin identification and loading
- number crunching (but maybe the new math abstractions will help here)
- generic interfaces and classes
- loading from serialized representations where I store the type so that I can deserialize with that type

Especially when having generic interfaces/classes. Especially in cases where I need to translate a Type into a class. Something like 

```cs
public interface IImage
{
  public int SizeX {get;}
  public int SizeY {get;}
}

public class Image<T>: IImage where T : struct
{
  public int SizeX {get;}
  public int SizeY {get;}
  public readonly T[] Data;

  public Image(int sizeX, int sizeY)
  {
    SizeX= sizeX;
    SizeY = sizeY;
    Data = new T[SizeX * SizeY);
  }

  public static IImage Create(Type t, int sizeX, int sizeY)
  {
    if (typeof(T) == typeof(float)) return new Image<float>(sizeX,sizeY);
    else if (typeof(T) == typeof(int)) return new Image<int>(sizeX,sizeY);
    else if (typeof(T) == typeof(uint)) return new Image<uint>(sizeX,sizeY);
    else if (typeof(T) == typeof(Half)) return new Image<Half>(sizeX,sizeY);
    else if (typeof(T) == typeof(double)) return new Image<double>(sizeX,sizeY);
...
  }
}
```

First of all the example does not extend itself when newer types/struct are introduced (e.g. for the new Half "type" I had to search through my whole codebase to find out where I needed to add the comparison). Sure in the example you could say do it with an Activator. But as we all know the Activator is sooo slow. Another question could be, why I don't write this:

```cs
var image = new Image<float>(512,512);
```

Because the type is stored as XML and I read the type from there. If I could write it like below this would solve all problems.

```cs
Type t = Type.GetType(typeString);
var image = new Image<t>(512,512);

```

Another example is this one (not easy to write, not easy to read, not easy to maintain)

```cs
public readonly ref struct ConvertedValue<TFrom, TTo> where TFrom:unmanaged where TTo:unmanaged 
    {
        private readonly Span<TFrom> InternalValue;

        public TTo this[int index]
        {
            [SkipLocalsInit]
            [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
            get
            {
                if ((typeof(TFrom)==typeof(byte)) && (typeof(TTo)==typeof(float))) 
                    return (TTo)(object)(float)(int)(object)InternalValue[index];
                else if ((typeof(TFrom) == typeof(sbyte)) && (typeof(TTo) == typeof(float)))
                    return (TTo)(object)(float)(sbyte)(object)InternalValue[index];
                else if ((typeof(TFrom) == typeof(ushort)) && (typeof(TTo) == typeof(float)))
                    return (TTo)(object)(float)(ushort)(object)InternalValue[index];
                else if ((typeof(TFrom) == typeof(short)) && (typeof(TTo) == typeof(float)))
                    return (TTo)(object)(float)(short)(object)InternalValue[index];

                throw new NotSupportException();
            }
        }

        public ConvertedValue(Span<TFrom> value)
        {
            InternalValue = value;
        }
    }
```

**Reply by HaloFour** [reply]
I'd probably emit expression trees representing invoking the constructor and compile that into a lambda and caching the result, that way you avoid the cost of `Activator.CreateInstance`.  That's probably the best that the compiler could do too.

**Reply by msedi** [reply]
@HaloFour. I didn't show that here, but that's what I already do. The problem is that I need to do this very often and writing expression trees is not OK but not my preferred approach, so I was just showing that there is room for improvement although maybe not on the high priority list ;-)

**Reply by ahdung** [reply]
@huoyaoyuan I'm just tired of using
`typeof(Foo).IsAssignableFrom(type) || typeof(Bar).IsAssignableFrom(type) || typeof(Baz).IsAssignableFrom(type) ...`
I hate `typeof` actually.

**Reply by ahdung** [reply]
Another case:
```c#
static bool IsNumericalType(Type type) =>
    type == typeof(byte)
    || type == typeof(short)
    || type == typeof(int)
    || type == typeof(long)
    || type == typeof(float)
    || type == typeof(double)
    || type == typeof(decimal)
    || type == typeof(sbyte)
    || type == typeof(ushort)
    || type == typeof(uint)
    || type == typeof(ulong)
    || type == typeof(byte?)
    || type == typeof(short?)
    || type == typeof(int?)
    || type == typeof(long?)
    || type == typeof(float?)
    || type == typeof(double?)
    || type == typeof(decimal?)
    || type == typeof(sbyte?)
    || type == typeof(ushort?)
    || type == typeof(uint?)
    || type == typeof(ulong?);

// in other place
if (IsNumericalType(column.DataType))
  column.Style.Alignment = Right;
```

**Reply by CyrusNajmabadi** [reply]
but you'll still have to write something effectively as complex even if we add a language feature here. You'll still need 20 clauses listed.  And you'll still write it over 20 lines. This is a case where they existing mechanism feels fine and a new language feature wouldn't help here with those sorts of patterns.

**Reply by CyrusNajmabadi** [reply]
Also, consider just unwrapping I'm the case of nullabele and just do 10 checks on that.

**Reply by ahdung** [reply]
@CyrusNajmabadi I agree that it will only bring a small benefit, seems so. But I still hope c# getting more and more beautiful and elegant.

**Comment by HaloFour** [comment]
I'll note that the first example is legal C# syntax which means something different.

**Reply by ahdung** [reply]
@HaloFour Yes it is, I just hope in the context of type matching, the instance of `System.Type` has different logic from other types.

**Reply by HaloFour** [reply]
@ahdung 

> I just hope in the context of type matching

In the context of type matching that variable is already of type `Type`, and that can't be changed without potentially breaking a lot of code.  I don't imagine that would be on the table.

As for generic type parameters, or perhaps a `typeof(...)` pattern, that at least would be new syntax and would have new meaning.  My question is whether or not such syntax actually moves the needle much in the examples provided above.  IMO, it doesn't, really.  These types of methods are pretty rare, so coming up with a new syntax and features specific to them seems a bit unnecessary.

**Comment by DavidArno** [comment]
The code examples here highlight the fact that there are two features that go hand in hand that are currently missing from C#'s pattern matching and generic handling abilities: being able to match against generic parameters and being able to dynamically invoke a generic method without using reflection. 

The simplest way to explain both of those is via expressing ahdung's `IsNumericalType` in the way I'd want to be able to write it:
```c#
static bool IsNumericalType<T>() => T switch {
    Nullable<var innerType> => IsNumericalType<innerType>(),
    byte or short or int or long or float or double or decimal or sbyte or ushort or uint or ulong => true,
    _ => false
};
```
Both features would be incredibly useful to me as I bump into this sort of problem often and have to invent various hacks and work-arounds to solve those problems. But if Cyrus's view is typical of the language team, they see such problems as so rare it's not worth the time of fixing them in the language and it's better to leave the small number of people affected inventing their various hacks and work-arounds. As I said before, I think it a huge shame.

**Reply by msedi** [reply]
@DavidArno:  

> they see such problems as so rare it's not worth the time of fixing them in the language

I think the language team does not really see this as a problem because in these "rare" cases they have the possibility to get around this asking the runtime to help e.g. [JitIntrinsic] or to introduce ByRef<T> that is not easily available or sometimes unavailable to the outside user. Vector.As<> is a typical example:

```cs
  /// <summary>
        /// Casts the given object to the specified type, performs no dynamic type checking.
        /// </summary>
        [Intrinsic]
        [NonVersionable]
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        [return: NotNullIfNotNull("value")]
        public static T As<T>(object? value) where T : class?
        {
            throw new PlatformNotSupportedException();
 
            // ldarg.0
            // ret
        }
```

I can speak for myself in my job. I'm sometimes also not seeing things customers complain because I created them and I'm more into this topic than anyone else.

That shouldn't be an excuse though...

**Reply by HaloFour** [reply]
It's a question of ROI.  I've written a _lot_ of reflection code over the years and still these two situations are pretty rare.  Creating a language feature that applies to a tiny percentage of the audience and that only reduces verbosity slightly isn't a good investment.  It'd take more keystrokes to implement than will likely ever be saved.

As for the ability to dynamically invoke a generic method, that does sound interesting, but that's still reflection as the compiler would effectively have to write exactly what you would write.  Given all generic and overload resolution would have to be bound at runtime that feels like something more appropriate to `dynamic`, and just as brittle.

**Reply by CyrusNajmabadi** [reply]
> and have to invent various hacks and work-arounds to solve those problems. 

Why do you need hacks? It seems trivial to write using standard .net mechanisms. Even if we had a lang feature it would likely compile down to the same. 

In the case of that method it literally seems quite trivial to write today using System.Type.

**Reply by CyrusNajmabadi** [reply]
> As I said before, I think it a huge shame.

You mentioned how rare this is. But you also think it's a huge shame.  It seems weird that something would simultaneously be both very uncommon, not 6 also hugely important.

**Reply by DavidArno** [reply]
@CyrusNajmabadi,

Apologies, I wasn't clear. I don't think use-cases for such functionality is rare at all. I was saying that you and the rest of the language team seem to have that view.

The problem is that to do the equivalent of `IsNumericalType<innerType>()` currently requires lots of boilerplate reflection code, which requires looking up how to do it, ploughing through arcane code in the examples etc. It's just not worth it. So I use completely different approaches to the problem.

It's a bit like pattern matching itself. Before the language supported it directly, there was a vanishingly small number of us that came up with fluent-syntax ways of simulating it. If you'd based the decision on whether to add pattern matching to the language on the number of people trying to simulate it, you'd never have bothered.

It would be interesting to know how often the pattern `if (typeof(T) == typeof(Xxxx))` exists in C# code as every single occurrence would benefit from expressing it as `T is Xxxx`. And having that then opens the way for people to express things that are currently done in other ways in that clearer and more succinct fashion.

You think the use-cases for this functionality are rare. I think you are wrong.

**Reply by HaloFour** [reply]
> Apologies, I wasn't clear. I don't think use-cases for such functionality is rare at all. I was saying that you and the rest of the language team seem to have that view.

Sounds like an assertion that one could quantify through an analysis of C# source on Github.

**Reply by siegfriedpammer** [reply]
https://grep.app/search?q=typeof%5Cs%2A%5C%28%5BA-z0-9%5D%2B%5C%29%5Cs%2A%5B%3D%21%5D%3D%5Cs%2Atypeof%5Cs%2A%5C%28%5BA-z0-9%5D%2B%5C%29&regexp=true

Not sure if that covers all repos, however it's a start. Apologies for any typos in the regex, I was writing this on my phone.

**Comment by CyrusNajmabadi** [comment]
> If you'd based the decision on whether to add pattern matching to the language on the number of people trying to simulate it, you'd never have bothered.

That's not true. We had ample cases of people needing to effectively do their own form of pattern matching with 'is' checks, property checks, comparisons, etc.  And the additional things we're doing in the matching space (like list patterns) are due to us seeing that as well. 

The code you're mentioning though seems far less common. It's much more of a niche case.
